pkg/util/kubernetes/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	if globalLeaderLeaseDuration == 0 {
+		globalLeaderLeaseDuration = defaultLeaderLeaseDuration
+	}
+	le.LeaseDuration = globalLeaderLeaseDuration
+	log.Debugf("LeaderLeaseDuration is %s", globalLeaderLeaseDuration.String())
+
+	le.coreClient, err = GetClient()
+	if err != nil {
+		log.Errorf("Not Able to set up a client for the Leader Election: %s", err)
+		return err
+	}
+
+	// check if we can get endpoints.
+	_, err = le.coreClient.Endpoints(metav1.NamespaceDefault).List(metav1.ListOptions{Limit: 1})
+	if err != nil {
+		log.Errorf("Cannot retrieve endpoints from the %s namespace: %s", metav1.NamespaceDefault, err)
+		return err
+	}
+
+	le.leaderElector, err = le.newElection(le.LeaseName, metav1.NamespaceDefault, le.LeaseDuration)
+	if err != nil {
+		log.Errorf("Could not initialize the Leader Election process: %s", err)
+		return err
+	}
+	log.Debug("Leader Engine successfully initialized")
+	return nil
+}
+
+func (le *LeaderEngine) startLeaderElection() {
+	log.Infof("Starting Leader Election process for %q ...", le.HolderIdentity)
+	go wait.Forever(le.leaderElector.Run, 0)
+}
+
+// EnsureLeaderElectionRuns start the Leader election process if not already started,
+// return nil if the process is effectively running
+func (le *LeaderEngine) EnsureLeaderElectionRuns() error {
+	var leaderIdentity string
+
+	le.once.Do(le.startLeaderElection)
+	timeoutDuration := clientTimeout * 2
+	timeout := time.After(timeoutDuration)
+	tick := time.NewTicker(time.Millisecond * 500)
+	for {
+		select {
+		case <-tick.C:
+			leaderIdentity = le.CurrentLeaderName()
+			if leaderIdentity != "" {
+				log.Infof("Leader Election run, current leader is %q", leaderIdentity)
+				return nil
+			}
+			log.Tracef("Leader identity is unset")
+
+		case <-timeout:
+			return fmt.Errorf("leader election still not running, timeout after %s", timeoutDuration.String())
+		}
+	}
+}
+
+// GetClient returns an official Kubernetes core v1 client
+func GetClient() (*corev1.CoreV1Client, error) {
+	var k8sConfig *rest.Config
+	var err error
+
+	cfgPath := config.Datadog.GetString("kubernetes_kubeconfig_path")
+	if cfgPath == "" {
+		k8sConfig, err = rest.InClusterConfig()
+		if err != nil {
+			log.Debug("Can't create a config for the official client from the service account's token: %s", err)
+			return nil, err
+		}
+	} else {
+		// use the current context in kubeconfig
+		k8sConfig, err = clientcmd.BuildConfigFromFlags("", cfgPath)
+		if err != nil {
+			log.Debug("Can't create a config for the official client from the configured path to the kubeconfig: %s, ", cfgPath, err)
+			return nil, err
+		}
+	}
+
+	k8sConfig.Timeout = clientTimeout
+	coreClient, err := corev1.NewForConfig(k8sConfig)
+
+	return coreClient, err
+}
+
+// CurrentLeaderName is the main interface that can be called to fetch the name of the current leader.
+func (le *LeaderEngine) CurrentLeaderName() string {
+	le.currentHolderMutex.RLock()
+	defer le.currentHolderMutex.RUnlock()
```

Comment:
not sure about the mutex use here, won't it be unlock before the value is read because it's in the return statement?
maybe you should have intermediary var